tests/test_json_to_parquet.py

```diff
@@ -0,0 +1,531 @@
+                            {
+                                "Name": "ArrayOfObjectsField",
+                                "Type": "array<struct<filename:string,timestamp:string>>"
+                            },
+                            {
+                                "Name": "ObjectField",
+                                "Type": "struct<filename:string,timestamp:string>"
+                            },
+                            {
+                                "Name": "export_end_date",
+                                "Type": "string"
+                            }
+                        ]
+                    },
+                    "Parameters": {
+                        "classification": "json",
+                        "compressionType": "none",
+                        "typeOfData": "file",
+                    }
+                }
+        )
+        return glue_table
+
+    @pytest.fixture(scope="class")
+    def glue_flat_table_location(self, glue_database_path, glue_flat_table_name):
+        glue_flat_table_location = os.path.join(
+                glue_database_path,
+                glue_flat_table_name.replace("_", "=", 1)
+            ) + "/"
+        return glue_flat_table_location
+
+    @pytest.fixture(scope="class")
+    def glue_flat_table(self, glue_database_name, glue_database_path,
+                        glue_flat_table_name, glue_flat_table_location):
+        glue_client = boto3.client("glue")
+        glue_table = glue_client.create_table(
+                DatabaseName=glue_database_name,
+                TableInput={
+                    "Name": glue_flat_table_name,
+                    "Description": "A table for pytest unit tests.",
+                    "Retention": 0,
+                    "TableType": "EXTERNAL_TABLE",
+                    "StorageDescriptor": {
+                        "Location": glue_flat_table_location,
+                        "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
+                        "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
+                        "Compressed": False,
+                        "StoredAsSubDirectories": False,
+                        "Columns": [
+                            {
+                                "Name": "GlobalKey",
+                                "Type": "string"
+                            },
+                            {
+                                "Name": "export_end_date",
+                                "Type": "string"
+                            }
+                        ]
+                    },
+                    "Parameters": {
+                        "classification": "json",
+                        "compressionType": "none",
+                        "typeOfData": "file",
+                    }
+                }
+        )
+        return glue_table
+
+    def upload_test_data_to_s3(self, path, s3_bucket, table_location):
```

Comment:
I think you could upload the test data with moto: https://docs.getmoto.org/en/latest/docs/services/s3.html and the rest of the functions can use that

Reply:
I'm uncertain if moto can do what I need it to do in these tests. I borrowed this method of testing wherein I don't mock some AWS resources from Bridge Downstream. Back then, Moto had even less support for Glue than it does now and I wanted to do some complicated things that I still don't fully understand -- particularly, updating the partitions and metadata in the Glue table so that when the JSON to Parquet job calls `glue_context.create_dynamic_frame.from_catalog` we actually get data. I know how to do that by running the crawler, but I spent a few days trying to do it manually a while back and couldn't figure out how to do it without running the crawler first. So it's still a bit of a mystery to me. And if I can't read the mocked tables then moto is useless.

I think a moto rewrite could be in the cards for the future but since we already have a working solution I'm reluctant to spend a day exploring what could be an unfruitful way to potentially shave a few minutes off test runtime.

Reply:
I think it might be that you have to merge `glue_database` and `test_glue_database` so they are in the same function.

Reply:
Yes, this seems to work. Thank you!

Reply:
Unfortunately that wasn't the hard part. Even if I mock all the AWS resources, all the actual work is being done with Dynamic Frames, meaning I still need to mock `GlueContext` (and `SparkContext`?) and `DynamicFrame`. The Context objects do their work within a Spark environment and I don't know what I can realistically mock. If I'm going to finish these tests by the end of this sprint I need to go with the current "dirty" approach rather than attempt a pure unit test style approach.